lib/xenops_server.ml

```diff
@@ -3282,10 +3289,14 @@ module VM = struct
       (fun () ->
         let id, md = parse_metadata s in
         (* We allow a higher-level toolstack to replace the metadata of a
-           running VM. Any changes will take place on next reboot. *)
+           running VM. Any changes will take place on next reboot.
+           The metadata update will be queued so that ongoing operations
+           do not see unexpected state changes. *)
         if DB.exists id then
-          debug "Overwriting VM metadata for VM: %s" id ;
-        import_metadata id md)
+          let _ = queue_operation_and_wait dbg id (Atomic (VM_import_metadata (id, md))) in
+          id
+        else
+          import_metadata id md)
```

Comment:
Can this immediate call change the metadata affect an ongoing vm creation operation?

Reply:
It is only done immediately if the VM is yet unknown to xenopsd (`DB.exists id = false`). In that case, xenopsd can't have any operations acting on that VM. "Creating" a VM, i.e. starting one, needs several calls to xenopsd: `VM.import_metadata` + `VM.start` + `VM.unpause`.